src/hooks/useOptionsBalance.ts

```diff
@@ -0,0 +1,35 @@
+import { useQuery } from "react-query";
+import { ethers } from "ethers";
+
+import { getKpiTokenContract } from "../utils";
+import { useConnection } from "./useConnection";
+import { infuraId } from "../config";
+
+export function useOptionsBalance(address?: string) {
+  const { provider, chainId } = useConnection();
+  const { data, ...others } = useQuery(
+    ["balance", address, provider?.connection, chainId],
+    () => getBalance(address!, provider, chainId),
+    {
+      enabled: address != null,
+    }
+  );
+  return {
+    balance: data ? ethers.utils.formatEther(data) : undefined,
+    ...others,
+  };
+}
+
+async function getBalance(
+  address: string,
+  web3Provider?: ethers.providers.Web3Provider,
+  chainId = 1
```

Comment:
It seems like it might be a bit of an anti-pattern to have this variable be defaulted across many functions:
- It could create subtle differences across environments. For instance, things may be consistent on mainnet, but be inconsistent on other networds.
- Are there actually cases where we expect to rely on this default? If so, how could that ever work on kovan, for instance?

Reply:
Let me add some context around this:
One thing we usually want from our dapps is to work without having to connect a wallet. When a user first loads the page, we don't have a preferred `chainId` or `provider`, so we have to fallback to something and the most sensible thing is to fall back to mainnet. 

If you assume this is the default behavior we want, then it makes sense to default `chainId` and `provider` in the hook definition.
As otherwise each time you consume it you would have to do something like this:

```javascript
function MyComp(){
// if the user hasn't connected their wallet, both of those are undefined
 const {provider, chainId} = useConnection()
 const _chainId = chainId ?? 1 
 const _provider = provider ?? someDefaultProvider
 const {balance} = useBalance(_provider, _chainId);
}
```

vs doing this

```javascript
function MyComp(){
// if the user hasn't connected their wallet, both of those are undefined
 const {provider, chainId} = useConnection()
 const {balance} = useBalance(provider, chainId);
}
```

Beside the savings in provider initializations (which could be saved as well in the first code block by adding a `useMemo` or a `useEffect`) I prefer the second code block as its easier to read and you can assume things will "just work" in your components, without having to think about the implementation details. 
Regarding inconsistency, it is consistent with the experience a user would get, and after all, after we connect our wallet, our UI would update to reflect the chain we are on

Reply:
hmm yea i remember this being a really complex issue. thanks for the context.